### PR TITLE
Adds the Grover Diffusion Operator Template

### PR DESCRIPTION
**Pull request by albi3ro**
This PR adds the Grover Diffusion Operator as a template.  

With this template, users can perform a Grover Search simply by the code :

```python
@qml.qnode(dev)
def GroverSearch(n=1):
    for wire in wires:
        qml.Hadamard(wire)
    for _ in range(n):
        oracle()
        qml.templates.GroverOperator(wires=wires)
    return qml.probs(wires)
```

Template documentation rendered [here](https://pennylane--1442.org.readthedocs.build/en/1442/code/api/pennylane.templates.subroutines.GroverOperator.html?highlight=grover#pennylane.templates.subroutines.GroverOperator)

### CODE REVIEW

File: pennylane/templates/subroutines/grover.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    r"""Performs the Grover Diffusion Operator.
+
+    .. math::
+
+        G = 2 |s \rangle \langle s | - I
+        = H^{\bigotimes n} \left( 2 |0\rangle \langle 0| - I \right) H^{\bigotimes n}
+
+
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:32]: Recommend to align the picture with the description; the diagram skips a couple of simplification steps that happen when you apply the identity `HZH = X`. One option is to have a new diagram just showing the circuit as described here, but then that wouldn't match the implementation. So maybe for clarity both versions can be shown, with an equals sign in the middle and brief mention of how you get one from the other.

In either case, I think it'd be good to mention that the open-circle notation means to control on 0 (or alternatively, have a layer of X gates).

File: pennylane/templates/subroutines/grover.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    .. code-block:: python
+
+        n_wires = 3
+        wires = list(range(n_wires))
+        dev = qml.device('default.qubit', wires=wires)
+
+    The Grover Diffusion Operator amplifies the magnitude of the component with
+    a negative phase.  For example, if we wanted to select out the :math:`|111\rangle`
+    state, we could use an oracle implementing a `CCZ` gate:
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:56]:
```suggestion
    a negative phase.  For example, if the solution to the search problem is the :math:`|111\rangle`
    state, we require an oracle that flips its phase; this could be implemented using a `CCZ` gate:
```

File: pennylane/templates/subroutines/grover.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    .. code-block:: python
+
+        def oracle():
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:60]: Out-of-scope, but we should probably implement `CCZ` at some point as an alias for this, I've tried to call it more than once :facepalm:

File: pennylane/templates/subroutines/grover.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            qml.Hadamard(wires[-1])
+            qml.Toffoli(wires=wires)
+            qml.Hadamard(wires[-1])
+
+    We can then implement the entire Grover Search Algorithm for ``n`` iterations:
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:65]:
```suggestion
    We can then implement the entire Grover Search Algorithm for ``n`` iterations by alternating calls to the oracle and the diffusion operator:
```

File: pennylane/templates/subroutines/grover.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        @qml.qnode(dev)
+        def GroverSearch(n=1):
+            for wire in wires:
+                qml.Hadamard(wire)
+
+            for _ in range(n):
+                oracle()
+                qml.templates.GroverOperator(wires=wires)
+            return qml.probs(wires)
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:77]: Nice! This looks super clean :grin:

File: pennylane/templates/subroutines/grover.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    """
+    num_params = 0
+    num_wires = AnyWires
+    par_domain = None
+
+    def __init__(self, wires=None, work_wires=None, do_queue=True, id=None):
+        self.work_wires = work_wires
+        super().__init__(wires=wires, do_queue=do_queue, id=id)
+
+    def expand(self):
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:95]: :100:

File: pennylane/templates/subroutines/grover.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    Args:
+        wires (Union[Wires, Sequence[int], or int]): the wires to apply to
+        work_wires (Union[Wires, Sequence[int], or int]): these auxiliary wire assist
+            in the decomposition of :class:`~.ops.qubit.MultiControlledX`.
```
[the remark this answers is not in the document]
Comment (glassnotes) [reply pennylane/templates/subroutines/grover.py:42]: Should we mention here how many wires would be needed? (Maybe not necessary since it links to the docs of `MultiControlledX` which contains those details?)

File: pennylane/templates/subroutines/grover.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                qml.templates.GroverOperator(wires=wires)
+            return qml.probs(wires)
+
+    >>> GroverSearch(n=1)
+    tensor([0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125,
+            0.78125], requires_grad=True)
+    >>> GroverSearch(n=2)
+    tensor([0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125,
+        0.0078125, 0.9453125], requires_grad=True)
+
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:85]: Briefly note that the largest probability is in the solution state, as expected.

File: tests/templates/test_subroutines/test_grover.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+"""
+Tests for the Grover Diffusion Operator template
+"""
+import pytest
+import numpy as np
+import pennylane as qml
+
+
+def test_work_wires():
+    """assert work wires get passed to MultiControlledX"""
```
Comment (glassnotes) [inline tests/templates/test_subroutines/test_grover.py:23]:
```suggestion
    """Assert work wires get passed to MultiControlledX"""
```

File: tests/templates/test_subroutines/test_grover.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    op = qml.templates.GroverOperator(wires=(0, 1), id="hello")
+
+    assert op.id == "hello"
+
+
+class TestDecomposition:
+    def test_expand(self):
+        """Tests correct expansion"""
+        wires = tuple(range(3))
```
Comment (glassnotes) [inline tests/templates/test_subroutines/test_grover.py:56]: Since the circuit has a predictable structure, perhaps this could be made a parametrized test that checks for a variety of different number of wires and label styles, rather than having two separate tests for integer vs. string labels.

File: pennylane/templates/subroutines/grover.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    """
+    num_params = 0
+    num_wires = AnyWires
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:88]: I don't think this affects the specification of `num_wires` itself, but special handling of a single input-wire case may need to be implemented; I don't think it would work for that (?)

**Review by glassnotes — CHANGES_REQUESTED** [review]
Thanks for implementing this @albi3ro ! This is a huge help for streamlining the implementation of Grover. The code looks good, mostly just minor comments for clarity of documentation.

File: pennylane/templates/subroutines/grover.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/grover.py:60]: Maybe just add a `MultiControlledZ` like the `MultiControlledX`.

File: .github/CHANGELOG.md [hunk 12]
```diff
@@ -2,6 +2,9 @@
 
 <h3>New features since last release</h3>
 
+* Grover Diffusion Operator template added.
+  [(#1442)](https://github.com/PennyLaneAI/pennylane/pull/1442)
+
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply .github/CHANGELOG.md:7]: Added.

File: pennylane/templates/subroutines/grover.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    >>> GroverSearch(num_iterations=1)
+    tensor([0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125,
+            0.78125], requires_grad=True)
+    >>> GroverSearch(num_iterations=2)
+    tensor([0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125,
+        0.0078125, 0.9453125], requires_grad=True)
+
+    We can see that the marked :math:`|111\rangle` state has the greatest probability amplitude.
+
+    Optimally, the oracle-operator pairing should be repeated :math:`\frac{\pi}{4}\sqrt{2^{n}}$ times.
```
Comment (rmoyard) [inline pennylane/templates/subroutines/grover.py:94]: The equation does not render well in the documentation.

File: pennylane/templates/subroutines/grover.py (hunk 11 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/grover.py:100]: I've added a check on the number of wires.

File: tests/templates/test_subroutines/test_grover.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+def test_id():
+    """Assert id keyword works"""
+
+    op = qml.templates.GroverOperator(wires=(0, 1), id="hello")
+
+    assert op.id == "hello"
+
+
+def decomposition(wires):
```
Comment (rmoyard) [inline tests/templates/test_subroutines/test_grover.py:61]: Is this used somewhere else?

**Review by rmoyard — COMMENTED** [review]
Great to have a simple way to implement Grover's search 💯

File: tests/templates/test_subroutines/test_grover.py (hunk 14 above)
Comment (albi3ro) [reply tests/templates/test_subroutines/test_grover.py:61]: oops... was in progress rewriting those tests than forgot about it.

File: pennylane/templates/subroutines/grover.py (hunk 13 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/grover.py]: start in markdown, end in latex... oops

File: pennylane/templates/subroutines/grover.py [hunk 15]
```diff
@@ -36,59 +41,68 @@ class GroverOperator(Operation):
...
     tensor([0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125,
             0.78125], requires_grad=True)
-    >>> GroverSearch(n=2)
+    >>> GroverSearch(num_iterations=2)
     tensor([0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125,
         0.0078125, 0.9453125], requires_grad=True)
 
+    We can see that the marked :math:`|111\rangle` state has the greatest probability amplitude.
+
+    Optimally, the oracle-operator pairing should be repeated :math:`\frac{\pi}{4}\sqrt{2^{n}}` times.
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:95]:
```suggestion
    Optimally, the oracle-operator pairing should be repeated :math:`\lceil \frac{\pi}{4}\sqrt{2^{n}} \rceil` times.
```

File: tests/templates/test_subroutines/test_grover.py [hunk 16]
```diff
@@ -50,65 +58,63 @@ def test_id():
...
+    qml.Hadamard,
+    qml.Hadamard,
+]
 
-        op = qml.templates.GroverOperator(wires=wires)
 
-        ops = op.expand().operations
+def decomposition_wires(wires):
+    wire_order = [
+        (wires[0],),
```
Comment (glassnotes) [inline tests/templates/test_subroutines/test_grover.py:74]: Are the `( ,)` necessary here? Instead could maybe just have `wire_order = [wires[0], ...]` for clarity, and in the check convert using `Wires` to get a proper wires object.

File: .github/CHANGELOG.md [hunk 17]
```diff
@@ -2,6 +2,49 @@
...
+      for wire in wires:
+          qml.Hadamard(wire)
+
+      for _ in range(num_iterations):
+          oracle()
+          qml.templates.GroverOperator(wires=wires)
+      return qml.probs(wires)
+  ```
+
+  We can see this circuit selects out the marked state:
```
Comment (glassnotes) [inline .github/CHANGELOG.md:37]:
```suggestion
  We can see this circuit yields the marked state with high probability:
```
(or something like this)

File: pennylane/templates/subroutines/grover.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
+
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    The open circles on the controlled gate indicates control on 0 instead of 1.
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:44]:
```suggestion
    The open circles on the controlled gate indicate control on 0 instead of 1.
```

File: pennylane/templates/subroutines/grover.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+    where :math:`n` is the number of wires, and :math:`|s\rangle` is the uniform superposition:
+
+    .. math::
+
+        |s\rangle = H^{\bigotimes n} |0\rangle =  \frac{1}{\sqrt{2^n}} \sum_{i=0}^{2^n} | i \rangle
+
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
+
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:38]:
```suggestion
    For this template, the operator is implemented with a layer of Hadamards, a layer of :math:`X`,
    followed by a multi-controlled :math:`Z` gate, then another layer of :math:`X` and Hadamards.
    This is expressed in a compact form by the circuit below: 

```

File: pennylane/templates/subroutines/grover.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    The open circles on the controlled gate indicates control on 0 instead of 1.
+    The ``Z`` gates on the last wire are a simplification of ``H X H``, where
+    where the last ``H`` gate converts the multi-controlled X gate into a
+    multi-controlled Z gate.
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:47]:
```suggestion
    where the last ``H`` gate converts the multi-controlled :math:`Z` gate into a
    multi-controlled :math:`X` gate.
```

File: pennylane/templates/subroutines/grover.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
+
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    The open circles on the controlled gate indicates control on 0 instead of 1.
+    The ``Z`` gates on the last wire are a simplification of ``H X H``, where
```
Comment (glassnotes) [inline pennylane/templates/subroutines/grover.py:45]:
```suggestion
    The ``Z`` gates on the last wire result from leveraging the circuit identity :math:`HXH = Z`, 
```

File: tests/templates/test_subroutines/test_grover.py [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+        (wires[2],),
+        wires,
+        (wires[2],),
+        (wires[0],),
+        (wires[1],),
+    ]
+    return wire_order
+
+
+@pytest.mark.parametrize("wires", ((0, 1, 2), ("a", "c", "b")))
```
Comment (glassnotes) [inline tests/templates/test_subroutines/test_grover.py:85]: Nice!

File: tests/templates/test_subroutines/test_grover.py (hunk 16 above)
Comment (albi3ro) [reply tests/templates/test_subroutines/test_grover.py:74]: `qml.wires.Wires(wires[0]) != wires[0]`, but `qml.wires.Wires(wires[0]) == (wires[0],  )`.

Maybe we should just improve the `__eq__` function for the wires class.

Comment (albi3ro) [reply tests/templates/test_subroutines/test_grover.py:74]: oh I get what you mean.  i'll do that.

File: pennylane/templates/subroutines/grover.py [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+
+    where :math:`n` is the number of wires, and :math:`|s\rangle` is the uniform superposition:
+
+    .. math::
+
+        |s\rangle = H^{\bigotimes n} |0\rangle =  \frac{1}{\sqrt{2^n}} \sum_{i=0}^{2^n-1} | i \rangle.
+
+    For this template, the operator is implemented with a layer of Hadamards, a layer of :math:`X`,
+    followed by a multi-controlled :math:`Z` gate, then another layer of :math:`X` and Hadamards.
+    This is expressed in a compact form by the circuit below:
```
Comment (albi3ro) [inline pennylane/templates/subroutines/grover.py:38]:
```suggestion
    This is expressed in a compact form by the circuit below:
```

File: pennylane/templates/subroutines/grover.py [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    followed by a multi-controlled :math:`Z` gate, then another layer of :math:`X` and Hadamards.
+    This is expressed in a compact form by the circuit below: 
+
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    The open circles on the controlled gate indicate control on 0 instead of 1.
+    The ``Z`` gates on the last wire result from leveraging the circuit identity :math:`HXH = Z`,
```
Comment (albi3ro) [inline pennylane/templates/subroutines/grover.py:46]:
```suggestion
    The ``Z`` gates on the last wire result from leveraging the circuit identity :math:`HXH = Z`,
```

File: .github/CHANGELOG.md [hunk 25]
```diff
@@ -2,6 +2,49 @@
...
+  dev = qml.device('default.qubit', wires=wires)
+
+  @qml.qnode(dev)
+  def GroverSearch(num_iterations=1):
+      for wire in wires:
+          qml.Hadamard(wire)
+
+      for _ in range(num_iterations):
+          oracle()
+          qml.templates.GroverOperator(wires=wires)
```
[the remark this answers is not in the document]
Comment (glassnotes) [reply .github/CHANGELOG.md:33]:
```suggestion
          qml.templates.GroverOperator(wires=wires)

```

**Review by glassnotes — APPROVED** [review]
Thanks @albi3ro ! Only remaining question for me is whether the font on the graphic can be changed to match the other templates; but otherwise this looks good to go on my end :tada:

**Comment by albi3ro** [comment]
@glassnotes  The text font might be changed by using a different pdf to svg converter, but I'm not sure the trouble figuring that out is worth the reward.
